### PR TITLE
feat: add `BitVec.(getMSbD, msb)_(add, sub)` and `BitVec.getLsbD_sub`

### PR DESCRIPTION
**Pull request by luisacicolini**
Since `getMsbD_add`, `getMsbD_sub`, `getLsbD_sub`, `msb_sub` , `msb_add` depend on `getLsbD_add` (which lives in`BitBlast.lean`) and on each other, I put all of these in `BitBlast.lean`.

### CODE REVIEW

File: src/Init/Data/BitVec/Bitblast.lean [hunk 1]
```diff
@@ -267,6 +267,38 @@ theorem add_eq_adc (w : Nat) (x y : BitVec w) : x + y = (adc x y false).snd := b
...
+theorem msb_add {w : Nat} {x y: BitVec w} :
+    (x + y).msb =
+      Bool.xor (getMsbD x 0) (Bool.xor (getMsbD y 0) (carry (w - 1) x y false)) := by
+  simp only [BitVec.msb, BitVec.getMsbD]
+  by_cases h : w ≤ 0
+  · simp [h, show w = 0 by omega]
+  · rw [getLsbD_add (x := x)]
+    simp [show w > 0 by omega]
+    omega
+theorem msb_sub {x y: BitVec w} :
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Bitblast.lean:284]: Can we get a newline in between the theorems here?

File: src/Init/Data/BitVec/Bitblast.lean [hunk 2]
```diff
@@ -267,6 +267,38 @@ theorem add_eq_adc (w : Nat) (x y : BitVec w) : x + y = (adc x y false).snd := b
 
 /-! ### add -/
 
+theorem getMsbD_add {i : Nat} {i_lt : i < w} {x y : BitVec w} :
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Bitblast.lean:270]: Is there a reason that these theorems are not `simp` unlike the previous ones?

File: src/Init/Data/BitVec/Bitblast.lean [hunk 3]
```diff
@@ -267,6 +267,38 @@ theorem add_eq_adc (w : Nat) (x y : BitVec w) : x + y = (adc x y false).snd := b
...
+theorem msb_sub {x y: BitVec w} :
+    (x - y).msb
+      = (x.getMsbD 0 ^^ ((~~~y + 1#w).getMsbD 0 ^^ carry (w - 1 - 0) x (~~~y + 1#w) false)) := by
+  simp [sub_toAdd, BitVec.neg_eq_not_add, msb_add]
+
+theorem getMsbD_sub {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    (x - y).getMsbD i =
+      (x.getMsbD i ^^ ((~~~y + 1).getMsbD i ^^ carry (w - 1 - i) x (~~~y + 1) false)) := by
+  rw [sub_toAdd, neg_eq_not_add, getMsbD_add]
+  rfl
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Bitblast.lean:293]: This looks like you are working on two goals? Can you focus them explicitly?

File: src/Init/Data/BitVec/Bitblast.lean (hunk 1 above)
Comment (hargoniX) [inline src/Init/Data/BitVec/Bitblast.lean:284]: I guess having a section for `sub` (see line 268) would be nice?

File: src/Init/Data/BitVec/Bitblast.lean (hunk 2 above)
Comment (luisacicolini) [reply src/Init/Data/BitVec/Bitblast.lean:270]: i assumed this was not `simp` since `getLsbD_add` isn't either and they have the same structure

Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Bitblast.lean:270]: Given the relative complex RHS I would feel this is indeed the safer choice.

**Review by tobiasgrosser — APPROVED** [review]

**Comment by luisacicolini** [comment]
awaiting-review

**Comment by hargoniX** [comment]
Having no `simp` there does sound reasonable if there is precedence for it and the rhs is so complicated yes. Though I do now wonder whether we *want* to have the simp on the `msb` variants? Is there a good reason for this?

**Comment by luisacicolini** [comment]
> Having no `simp` there does sound reasonable if there is precedence for it and the rhs is so complicated yes. Though I do now wonder whether we _want_ to have the simp on the `msb` variants? Is there a good reason for this?

no good reason. I guess I initially thought it could be `simp` but it is not in fact, so I removed it

**Pull request merged**